File: 04_Stack/03_Evaluate_RPN/main.py
class Solution:
    def evalRPN(self, tokens):
        stack = []

        for token in tokens:
            if token in "+-*/":
                b = stack.pop()
                a = stack.pop()
                if token == "+":
                    stack.append(a + b)
                elif token == "-":
                    stack.append(a - b)
                elif token == "*":
                    stack.append(a * b)
                elif token == "/":
                    # Python division truncates towards negative infinity by default, 
                    # so we need to use int() to ensure truncation towards zero.
                    stack.append(int(a / b))
            else:
                stack.append(int(token))

        return stack[0]

# Example usage
if __name__ == "__main__":
    sol = Solution()
    tokens = ["1", "2", "+", "3", "*", "4", "-"]
    result = sol.evalRPN(tokens)
    print(f"Result: {result}")  # Output: 5

File: 04_Stack/03_Evaluate_RPN/main.cpp
#include <vector>
#include <string>
#include <stack>
#include <stdexcept>

using namespace std;

class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> st;

        for (const string& token : tokens) {
            if (token == "+" || token == "-" || token == "*" || token == "/") {
                int b = st.top();
                st.pop();
                int a = st.top();
                st.pop();
                if (token == "+") st.push(a + b);
                if (token == "-") st.push(a - b);
                if (token == "*") st.push(a * b);
                if (token == "/") st.push(a / b);
            } else {
                st.push(stoi(token));
            }
        }

        return st.top();
    }
};

// Example usage
int main() {
    Solution sol;
    vector<string> tokens = {"1", "2", "+", "3", "*", "4", "-"};
    int result = sol.evalRPN(tokens);
    printf("Result: %d\n", result);  // Output: 5
    return 0;
}
